compiler/rustc_type_ir/src/elaborate.rs

```diff
@@ -79,20 +80,55 @@ pub fn elaborate<I: Interner, O: Elaboratable<I>>(
 ) -> Elaborator<I, O> {
     let mut elaborator =
         Elaborator { cx, stack: Vec::new(), visited: HashSet::default(), mode: Filter::All };
-    elaborator.extend_deduped(obligations);
+    elaborator.extend_deduped(None, obligations);
     elaborator
 }
 
 impl<I: Interner, O: Elaboratable<I>> Elaborator<I, O> {
-    fn extend_deduped(&mut self, obligations: impl IntoIterator<Item = O>) {
+    /// Adds `obligations` to the stack. `current_clause` is the clause which was elaborated to
+    /// produce these obligations.
+    fn extend_deduped(
+        &mut self,
+        current_clause: Option<I::Clause>,
+        obligations: impl IntoIterator<Item = O>,
+    ) {
         // Only keep those bounds that we haven't already seen.
         // This is necessary to prevent infinite recursion in some
         // cases. One common case is when people define
         // `trait Sized: Sized { }` rather than `trait Sized { }`.
         self.stack.extend(
-            obligations.into_iter().filter(|o| {
-                self.visited.insert(self.cx.anonymize_bound_vars(o.predicate().kind()))
-            }),
+            obligations
+                .into_iter()
+                .filter(|o| self.visited.insert(self.cx.anonymize_bound_vars(o.predicate().kind())))
+                .filter(|o| {
+                    let Some(current_clause) = current_clause else {
+                        return true;
+                    };
```

Comment:
Kinda weird that this is run every loop, but I do see why the alternatives would just be more roundabout

Reply:
Yeah, I had initially implemented it in the `elaborate` function but `extend_deduped` made more sense in the end.